crates/lsp/src/header_reader.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    io, mem,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{AsyncBufRead, Future};
+
+const HEADER_DELIMITER: &'static [u8; 4] = b"\r\n\r\n";
+
+pub fn read_headers<'a, R>(
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+) -> impl Future<Output = io::Result<usize>> + 'a
+where
+    R: AsyncBufRead + ?Sized + Unpin,
+{
+    HeaderReader {
+        reader,
+        buf,
+        read: 0,
+    }
+}
+
+struct HeaderReader<'a, R: ?Sized> {
+    reader: &'a mut R,
+    buf: &'a mut Vec<u8>,
+    read: usize,
+}
+
+impl<'a, R: ?Sized> HeaderReader<'a, R> {
+    fn read_headers<C: AsyncBufRead + ?Sized>(
```

Comment:
`read_headers` implementation assumes that headers could come in chunks - is that the correct assumption? If that's the case, we may want to add more sophisticated tests, ensuring that all code paths are exercised correctly (some guidance is welcome).